rbd/rbd_test.go

```diff
@@ -1259,3 +1323,97 @@ func TestClosedImage(t *testing.T) {
 	conn.DeletePool(poolname)
 	conn.Shutdown()
 }
+
+func TestOpenImage(t *testing.T) {
+	conn := radosConnect(t)
+
+	poolname := GetUUID()
+	err := conn.MakePool(poolname)
+	assert.NoError(t, err)
+
+	ioctx, err := conn.OpenIOContext(poolname)
+	require.NoError(t, err)
+
+	name := GetUUID()
+
+	_, err = OpenImage(ioctx, name, NoSnapshot)
+	assert.Error(t, err)
+
+	err = quickCreate(ioctx, name, testImageSize, testImageOrder)
+	assert.NoError(t, err)
+
+	oImage, err := OpenImage(ioctx, name, NoSnapshot)
+	assert.NoError(t, err)
+	assert.Equal(t, name, oImage.name)
+	err = oImage.Close()
+	assert.NoError(t, err)
+
+	// open read-only
+	oImage, err = OpenImageReadOnly(ioctx, name, NoSnapshot)
+	assert.NoError(t, err)
+
+	bytes_in := []byte("input data")
+	_, err = oImage.Write(bytes_in)
+	// writing should fail in read-only mode
+	assert.Error(t, err)
+
+	err = oImage.Close()
+	assert.NoError(t, err)
+
+	err = oImage.Remove()
+	assert.NoError(t, err)
+
+	_, err = OpenImageReadOnly(ioctx, name, NoSnapshot)
+	assert.Error(t, err)
+
+	ioctx.Destroy()
+	conn.DeletePool(poolname)
+	conn.Shutdown()
+}
+
+func TestRemoveImage(t *testing.T) {
+	conn := radosConnect(t)
+
+	poolname := GetUUID()
+	err := conn.MakePool(poolname)
+	assert.NoError(t, err)
+
+	ioctx, err := conn.OpenIOContext(poolname)
+	require.NoError(t, err)
+
+	// trying to remove a non-existent image is an error
+	err = RemoveImage(ioctx, "bananarama")
+	require.Error(t, err)
+
+	// create and then remove an image
+	name := GetUUID()
+	err = CreateImage(ioctx, name, testImageSize, NewRbdImageOptions())
```

Comment:
NewRbdImageOptions() allocates something through librbd. It needs a call to Destroy() too. Is it not possible to pass `nil`?

Reply:
Oops, good catch thanks. I had a bunch that were not calling destroy. Now updated.
It's not possible to pass nil currently. You added a check for that when it was still called Create4.